internal/errors.go

```diff
@@ -0,0 +1,67 @@
+package internal
+
+import (
+	"bytes"
+	"fmt"
+	"strings"
+
+	"github.com/cilium/ebpf/internal/unix"
+	"github.com/pkg/errors"
+)
+
+// ErrorWithLog returns an error that includes logs from the
+// kernel verifier.
+//
+// logErr should be the error returned by the syscall that generated
+// the log. It is used to check for truncation of the output.
+func ErrorWithLog(err error, log []byte, logErr error) error {
+	detail := strings.TrimRight(CString(log), "\t\r\n ")
+	if errors.Cause(logErr) == unix.ENOSPC {
+		detail += " (truncated...)"
+	}
+
+	// The most specific information is usually at the end of
+	// the log. Try to find the last line and try to use that
+	// as the summary.
+	var summary string
+	if pos := strings.LastIndexByte(detail, '\n'); pos > 0 {
```

Comment:
I haven't checked. My reasoning is this: since the log always appends to the end of the buffer, the most relevant information is by definition at the end. I can't really see a reason why this would be turned around.

I think the worst case is an error with the debug info spread over multiple lines, in which case this would break. When I tried it worked well, but I can remove it if you prefer.

Reply:
I don't feel super strongly about this, but I'm for removing it. It feels that at some point we'll mangle the verifier log and make it more confusing to interpret.

Reply:
This is what it looks like:
```
        btf_test.go:79: Can't load BTF: invalid argument: [26] DATASEC .maps size=0 vlen=1 size == 0
            magic: 0xeb9f
            version: 1
            flags: 0x0
            hdr_len: 24
            type_off: 0
            type_len: 608
            str_off: 608
            str_len: 589
            btf_total_size: 1221
            [1] STRUCT (anon) size=40 vlen=5
            	type type_id=2 bits_offset=0
            	key type_id=6 bits_offset=64
            	value type_id=6 bits_offset=128
            	max_entries type_id=2 bits_offset=192
            	map_flags type_id=2 bits_offset=256
            [2] PTR (anon) type_id=4
            [3] INT int size=4 bits_offset=0 nr_bits=32 encoding=SIGNED
            [4] ARRAY (anon) type_id=3 index_type_id=5 nr_elems=1
            [5] INT __ARRAY_SIZE_TYPE__ size=4 bits_offset=0 nr_bits=32 encoding=(none)
            [6] PTR (anon) type_id=7
            [7] TYPEDEF uint32_t type_id=8
            [8] INT unsigned int size=4 bits_offset=0 nr_bits=32 encoding=(none)
            [9] VAR btf_map type_id=1 linkage=1
            [10] FUNC_PROTO (anon) return=3 args=(3 arg)
            [11] FUNC helper_func2 type_id=10
            [12] FUNC_PROTO (anon) return=3 args=(3 arg)
            [13] FUNC helper_func type_id=12
            [14] FUNC_PROTO (anon) return=3 args=(void)
            [15] FUNC xdp_prog type_id=14
            [16] FUNC_PROTO (anon) return=3 args=(void)
            [17] FUNC no_relocation type_id=16
            [18] INT char size=1 bits_offset=0 nr_bits=8 encoding=SIGNED
            [19] ARRAY (anon) type_id=18 index_type_id=5 nr_elems=4
            [20] VAR __license type_id=19 linkage=1
            [21] STRUCT bpf_map_def size=20 vlen=5
            	type type_id=8 bits_offset=0
            	key_size type_id=8 bits_offset=32
            	value_size type_id=8 bits_offset=64
            	max_entries type_id=8 bits_offset=96
            	map_flags type_id=8 bits_offset=128
            [22] VAR hash_map type_id=21 linkage=1
            [23] VAR hash_map2 type_id=21 linkage=1
            [24] VAR array_of_hash_map type_id=21 linkage=1
            [25] VAR hash_of_hash_map type_id=21 linkage=1
            [26] DATASEC .maps size=0 vlen=1 size == 0
```

Do you still think we're likely to garble this output?

Reply:
I got rid of this, it doesn't do the right thing for verifier logs. Good call!